File: apps/android/app/src/main/java/chat/vostok/android/core/multiserver/ServerStore.kt
package chat.vostok.android.core.multiserver

import android.content.Context
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import chat.vostok.android.BuildConfig
import chat.vostok.android.core.storage.SessionStore
import java.io.File
import java.time.Instant

/**
 * Persists the server list to a JSON file in the app's internal storage.
 * Handles migration from legacy single-server auth (SessionStore + BuildConfig).
 */
class ServerStore(private val context: Context) {
    private val moshi = Moshi.Builder().build()
    private val listType = Types.newParameterizedType(List::class.java, ServerEntry::class.java)
    private val adapter = moshi.adapter<List<ServerEntry>>(listType).indent("  ")
    private val file: File get() = File(context.filesDir, FILE_NAME)
    private val lock = Any()

    fun load(): List<ServerEntry> = synchronized(lock) {
        try {
            if (!file.exists()) return emptyList()
            val json = file.readText()
            adapter.fromJson(json)?.sortedBy { it.sortOrder } ?: emptyList()
        } catch (e: Exception) {
            android.util.Log.w("ServerStore", "Failed to load servers: $e")
            emptyList()
        }
    }

    fun save(servers: List<ServerEntry>): Unit = synchronized(lock) {
        try {
            file.writeText(adapter.toJson(servers))
        } catch (e: Exception) {
            android.util.Log.e("ServerStore", "Failed to save servers: $e")
        }
    }

    fun add(entry: ServerEntry): List<ServerEntry> = synchronized(lock) {
        val servers = load().toMutableList()
        val withOrder = entry.copy(sortOrder = servers.size)
        servers.add(withOrder)
        save(servers)
        servers
    }

    fun update(entry: ServerEntry): List<ServerEntry> = synchronized(lock) {
        val servers = load().toMutableList()
        val idx = servers.indexOfFirst { it.id == entry.id }
        if (idx >= 0) servers[idx] = entry
        save(servers)
        servers
    }

    fun remove(id: String): List<ServerEntry> = synchronized(lock) {
        val servers = load().filter { it.id != id }
        val reordered = servers.mapIndexed { i, s -> s.copy(sortOrder = i) }
        save(reordered)
        reordered
    }

    /**
     * Migrates from single-server (SessionStore + BuildConfig) to multi-server.
     * Called once on first launch after the multi-server update.
     */
    fun migrateFromLegacySingleServer(sessionStore: SessionStore): List<ServerEntry> {
        val existing = load()
        if (existing.isNotEmpty()) return existing

        val legacySession = sessionStore.load() ?: return emptyList()

        val entry = ServerEntry.create(
            url = BuildConfig.VOSTOK_BASE_URL,
            label = BuildConfig.VOSTOK_INSTANCE_LABEL.ifBlank { "My Server" },
            color = "#FF5500",
        ).copy(
            auth = ServerAuth(
                token = legacySession.token,
                userId = legacySession.userId,
                username = legacySession.username ?: "",
                deviceId = legacySession.deviceId,
            ),
            device = ServerDevice(
                deviceId = legacySession.deviceId,
                identityPublicKey = "",
            ),
            lastConnectedAt = Instant.now().toString(),
        )

        val servers = listOf(entry)
        save(servers)
        return servers
    }

    companion object {
        private const val FILE_NAME = "vostok.servers.json"
    }
}

File: apps/ios/Vostok/Core/MultiServer/ServerStore.swift
import Foundation

/// Persists the server list to disk as JSON in the app support directory.
/// Handles migration from legacy single-server auth (KeychainSessionStore + Environment.plist).
final class ServerStore {
    private static let fileName = "vostok.servers.json"
    private let fileURL: URL
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    init() {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dir = base.appendingPathComponent("Vostok", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.fileURL = dir.appendingPathComponent(Self.fileName)
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
    }

    // MARK: – CRUD

    func load() -> [ServerEntry] {
        guard let data = try? Data(contentsOf: fileURL),
              let servers = try? decoder.decode([ServerEntry].self, from: data)
        else {
            return []
        }
        return servers.sorted(by: { $0.sortOrder < $1.sortOrder })
    }

    func save(_ servers: [ServerEntry]) {
        guard let data = try? encoder.encode(servers) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }

    func add(_ server: ServerEntry) -> [ServerEntry] {
        var servers = load()
        var entry = server
        entry.sortOrder = servers.count
        servers.append(entry)
        save(servers)
        return servers
    }

    func update(_ server: ServerEntry) -> [ServerEntry] {
        var servers = load()
        if let index = servers.firstIndex(where: { $0.id == server.id }) {
            servers[index] = server
        }
        save(servers)
        return servers
    }

    func remove(id: String) -> [ServerEntry] {
        var servers = load()
        servers.removeAll(where: { $0.id == id })
        for i in servers.indices {
            servers[i].sortOrder = i
        }
        save(servers)
        return servers
    }

    // MARK: – Legacy migration

    /// Migrates from single-server (KeychainSessionStore + Environment.plist) to multi-server.
    /// Called once on first launch after the multi-server update.
    /// Returns the migrated server list (or empty if nothing to migrate).
    func migrateFromLegacySingleServer(environment: AppEnvironment) -> [ServerEntry] {
        let existing = load()
        guard existing.isEmpty else { return existing }

        // Check if there's a legacy session in the Keychain
        guard let legacySession = KeychainSessionStore.shared.fetch() else {
            return []
        }

        // Create the first ServerEntry from the legacy data
        var server = ServerEntry.create(
            url: environment.baseURL.absoluteString,
            label: environment.instanceLabel.isEmpty ? "My Server" : environment.instanceLabel,
            color: "#FF5500"
        )

        server.auth = ServerAuth(
            token: legacySession.token,
            userID: legacySession.userID,
            username: legacySession.username,
            deviceID: legacySession.deviceID
        )

        server.device = ServerDevice(
            deviceID: legacySession.deviceID,
            identityPublicKey: ""
        )

        server.lastConnectedAt = ISO8601DateFormatter().string(from: Date())

        let servers = [server]
        save(servers)

        return servers
    }
}
